torch/csrc/jit/script/convert_to_ssa.cpp

```diff
@@ -0,0 +1,249 @@
+      }
+    }
+
+    // Following the same logic as emitIfElseBlocks in compiler.cpp,
+    // we emit a node output if the variable is defined in each block
+    // and the types of each block can be unified
+
+    for (const auto& x : mutated_variables) {
+      auto true_type = true_vars->findInAnyFrame(x);
+      auto false_type = false_vars->findInAnyFrame(x);
+      auto unified = unifyTypes(true_type, false_type);
+      if (!unified) {
+        continue;
+      }
+
+      addBlockOutput(true_block, true_type, x);
+      addBlockOutput(false_block, false_type, x);
+      addNodeOutput(n, *unified, x);
+    }
+  }
+
+  // loop_carried_outputs* = Loop(max_trip_count, start_condition,
+  //                              loop_carried_inputs*)
+  //                    block0(loop_counter, loop_carried_block*) {
+  //                       <body>
+  //                       -> (continue_condition, loop_carried_block_outputs*)
+  //                    }
+  // all loop_carried_... lists are the same length and represent the value of
+  // loop-carried variables whose definitions are updated as the loop executes
+  // in a way that ensure single static assignment.
+  void addLoopOutputs(Node* n) {
+    auto body_block = n->blocks().at(0);
+    auto loop_vars = addControlFlowOutputs(body_block);
+    for (const auto& name : loop_vars->definedVariables()) {
+      // we require that the variable is defined outside the loop to be emitted,
+      // and we do not refine the type of the parent variable since the loop may
+      // not be entered.
+      auto parent_type = environment_stack->findInAnyFrame(name);
+      if (!parent_type) {
+        continue;
+      }
+
+      // Insert a store at the beginning of the loop block, so that all
+      // loads of the variable will use the loop carried value
+      addNodeInput(n, parent_type, name);
+      addBlockInput(body_block, parent_type, name);
+      addBlockOutput(body_block, parent_type, name);
+      addNodeOutput(n, parent_type, name);
+    }
+
+    // loop continue expression should go after the loop carried outputs
+    auto loop_condition = body_block->outputs().at(0)->node();
+    AT_ASSERT(loop_condition->kind() == prim::LoopCondition);
+    loop_condition->moveBefore(body_block->return_node());
+    auto loop_condition_block = loop_condition->blocks().at(0);
+    for (auto it = loop_condition_block->nodes().begin();
+         it != loop_condition_block->nodes().end();) {
+      auto block_node = *it++;
+      block_node->moveBefore(loop_condition);
+    }
+
+    for (Node* n : loop_condition_block->nodes()) {
+      n->moveBefore(loop_condition);
+    }
+    body_block->eraseOutput(0);
+    body_block->insertOutput(0, loop_condition_block->outputs().at(0));
+    loop_condition->destroy();
+  }
+
+  std::shared_ptr<TypeEnvironment> addControlFlowOutputs(Block* block) {
+    pushFrame(block);
+    for (auto it = block->nodes().begin(); it != block->nodes().end(); it++) {
```

Comment:
`for(Node*node :block->nodes()`

Reply:
prim::If and prim::Loop are adding new nodes to the block... will try changing it